Fix inherited C# scene not inheriting parent's fields

kappa54m opened the pull request:
This fixes issue #36480 (which is the same bug as #37581).
The parent properties of a C# class were not fetched properly because it relies on CSharpInstance::get_property_list method, which fetches the field information from CSharpInstance::script->member_info, and CSharpScript::member_info is updated in methods CSharpScript::_update_exports and CSharpScript::_update_member_info_no_exports, the latter not declared if TOOLS_ENABLED is not set. The update to CSharpScript::member_info is not executed in the former method if TOOLS_ENABLED is not set, and by removing the restriction CSharpScript::member_info is correctly set up when method CSharpScript::_update_exports is called from CSharpScript::reload.

*Bugsquad edit:* Fixes #36480, fixes #37581.

kuruk-mm committed:
GDScript: Add join method on String
Docs: Add join on String.xml

akien-mga committed:
Style: Add missing copyright headers

akien-mga committed:
Style: clang-format: Disable AllowShortIfStatementsOnASingleLine

Part of #33027, also discussed in #29848.

Enforcing the use of brackets even on single line statements would be
preferred, but `clang-format` doesn't have this functionality yet.

akien-mga committed:
Style: clang-format: Disable AllowShortCaseLabelsOnASingleLine

Part of #33027.

Commit:
Fix inherited C# scene not inheriting parent's fields

When a child scene inherits a parent scene with a C# root node, the
parent scene's export variables appear to assume values set in the
parent scene, in the child scene's Inspector. However, when the child
scene is played, the parent scene's export variables assume default
values.
When a node is created, it inherits its parent C# script's fields from
the map CSharpScriptInstance::script->member_info. However this map was
not initialized outside the editor, and this commit ensured it is. This
fixes issues #36480 and #37581.

Commit:
Fix inherited C# scene not inheriting parent's fields

When a child scene inherits a parent scene with a C# root node, the
parent scene's export variables appear to assume values set in the
parent scene, in the child scene's Inspector. However, when the child
scene is played, the parent scene's export variables assume default
values.
When a node is created, it inherits its parent C# script's fields from
the map CSharpScriptInstance::script->member_info. However this map was
not initialized outside the editor, and this commit ensured it is. This
fixes issues #36480 and #37581.

pepegadeveloper123 commented:
Wait I think this might not compile if TOOLS_ENABLED is not set. Will work on this right now.

neikeq commented:
This would also fix #18293, but as I [commented](https://github.com/godotengine/godot/issues/18293#issuecomment-577318650) we need to address the issues #27014 fixed.

One possibility could be to analyze the source to get the initial values of fields and properties without having to call the constructor. This would be easy with the upcoming C# [source generators](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/), but is already possible in other ways. However, constructors can alter the initial values after initializing them, so we need to make sure that wouldn't cause problems.

akien-mga committed:
Merge pull request #38621 from akien-mga/stylé-comme-jamais

Style: clang-format: Disable if statements and case labels on single line

akien-mga committed:
Revert "Renamed plane's d to distance"

This reverts commit ec7b481170dcd6a7b4cf0e6c1221e204ff7945f3.

This was wrong, `d` is not a distance but the `d` constant in the
parametric equation `ax + by + cz = d` describing the plane.

akien-mga committed:
Merge pull request #38548 from kuruk-mm/add_string_join_to_gdscript

GDScript: Add join method on String

Commit:
Inherited scene bug & compile without TOOLS_ENABLED

The previous commit: Fix inherited C# scene not inheriting parent's
fields, did not account for variables that would be undefined with
TOOLS_ENABLED not set resulting in a compilation failure.
This fix will allow compilation with or without TOOLS_ENABLED option and
it accounts for #27014 by invoking the constructors only inside the
editor.

Commit:
Inherited scene bug merge conflict fix
